substrate/client/transaction-pool/src/fork_aware_txpool/tx_mem_pool.rs

```diff
@@ -604,29 +610,79 @@ where
 				.map(move |validation_result| {
 					xt.validated_at
 						.store(finalized_block.number.into().as_u64(), atomic::Ordering::Relaxed);
-					(xt_hash, validation_result)
+					(xt_hash, xt, validation_result)
 				})
 		});
 		let validation_results = futures::future::join_all(validations_futures).await;
 		let input_len = validation_results.len();
 
 		let duration = start.elapsed();
-
-		let invalid_hashes = validation_results
+		let mut revalidated = indexmap::IndexMap::new();
+		let invalid_txs = validation_results
 			.into_iter()
-			.filter_map(|(tx_hash, validation_result)| match validation_result {
-				Ok(Ok(_)) |
+			.filter_map(|(tx_hash, xt, validation_result)| match validation_result {
+				Ok(Ok(validity)) => {
+					revalidated.insert(
+						tx_hash,
+						ValidatedTransaction::valid_at(
+							finalized_block.number.into().as_u64(),
+							tx_hash,
+							xt.source.clone(),
+							xt.tx().clone(),
+							self.api.hash_and_length(&xt.tx()).1,
+							validity,
+						),
+					);
+					None
+				},
 				Ok(Err(TransactionValidityError::Invalid(InvalidTransaction::Future))) => None,
-				Err(_) |
-				Ok(Err(TransactionValidityError::Unknown(_))) |
-				Ok(Err(TransactionValidityError::Invalid(_))) => {
+				Err(ref error) => {
+					trace!(
+						target: LOG_TARGET,
+						?tx_hash,
+						?validation_result,
+						?error,
+						"mempool::revalidate_inner error during revalidation"
+					);
+					Some(ValidatedTransaction::Invalid(
+						tx_hash,
+						ChainApi::Error::from(
+							sc_transaction_pool_api::error::Error::InvalidTransaction(
+								sp_runtime::transaction_validity::InvalidTransaction::Custom(
+									InvalidTransactionCustomCode::RuntimeApiPrerequisitesFailure
+										as u8,
+								),
+							),
+						),
+					))
```

Comment:
This stuff is really not what should be recorded in a metric.

Reply:
Ok, I dropped the InvalidTransaction::Custom usage across the PR. Hope this looks better. LMK what you think.